src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -785,4 +769,40 @@ public void visitSimpleChunks(@NonNull SimpleChunkVisitor visitor, @NonNull Chun
         }
     }
 
+    /**
+     * Like {@link FlowNode#iterateEnclosingBlocks}, but if the original node is a {@link BlockEndNode}, its
+     * corresponding {@link BlockStartNode} is the first element in the iterator.
```

Comment:
This is kind of interesting to me, because I have been doing a lot of flow graph traversal lately and I have found that this behavior is what I typically want vs plain `iterateEnclosingBlocks`, so it was surprising to see the default behavior causing problems here. Also sometimes, if the node is itself is a `BlockStartNode`, I want to include it in the iteration as well. IDK if that matters here though - I think you would have to have a parallel branch start node in `FlowExecution.getCurrentHeads`, so it would only matter very temporarily.